src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs

```diff
@@ -0,0 +1,270 @@
+								        continue;
+
+					        //if this returns false, this means either timeout or canceled token is activated                    
+							while (_waitForChanges.Wait(_minimalHeartbeatInterval, _cts.Token) == false)
+						        SendHeartbeat(stream);
+				        }
+			        }
+		        }
+	        }
+	        catch (OperationCanceledException)
+	        {
+				if (_log.IsInfoEnabled)
+					_log.Info($"Operation canceled on replication thread ({FromToString}). Stopped the thread.");
+	        }
+            catch (Exception e)
+            { 
+				if(_log.IsInfoEnabled)
+					_log.Info($"Unexpected exception occured on replication thread ({FromToString}). Stopped the thread.", e);
+                Failed?.Invoke(this, e);
+            }
+        }
+
+	    private string FromToString => $"from {_database.ResourceName} to {_destination.Database} at {_destination.Url}";
+
+		private void SendHeartbeat(NetworkStream stream)
+	    {
+		    try
+		    {
+			    using (var writer = new BlittableJsonTextWriter(_context, stream))
+				    writer.WriteObjectOrdered(_heartbeatMessage);
+		    }
+			catch (Exception e)
+			{
+				if (_log.IsInfoEnabled)
+					_log.Info($"Sending heartbeat failed. ({_database.Name})", e);
+				throw;
+			}
+		}
+
+		private bool ExecuteReplicationOnce(BlittableJsonTextWriter writer, NetworkStream stream)
+	    {
+		    //just for shorter code
+		    var documentStorage = _database.DocumentsStorage;
+		    using (_context.OpenReadTransaction())
+		    {
+			    //TODO: make replication batch size configurable
+			    //also, perhaps there should be timers/heuristics
+			    //that would dynamically resize batch size
+			    var replicationBatch =
+				    documentStorage
+					    .GetDocumentsAfter(_context, _lastSentEtag, 0, 1024)
+					    .Where(x => !x.Key.ToString().StartsWith("Raven/"))
+					    .ToList();
+
+			    //the filtering here will need to be reworked -> it is not efficient
+			    //TODO: do not forget to make version of GetDocumentsAfter with a prefix filter	
+			    //alternatively 1 -> create efficient StartsWith() for LazyString				
+			    //alternatively 2 -> create a "filter system" that would abstract the logic -> what documents 
+			    //should and should not be replicated
+			    if (replicationBatch.Count == 0)
+				    return false;
+
+			    _cts.Token.ThrowIfCancellationRequested();
+
+			    SendDocuments(writer, stream, replicationBatch);
+			    return true;
+		    }
+	    }
+
+	    private void SendDocuments(BlittableJsonTextWriter writer, NetworkStream stream, IEnumerable<Document> docs)
+	    {
+		    if (docs == null) //precaution, should never happen
+				throw new ArgumentNullException(nameof(docs));
+
+		    if (_log.IsInfoEnabled)
+			    _log.Info($"Starting sending replication batch ({_database.Name})");
+
+		    var sw = Stopwatch.StartNew();
+		    writer.WriteStartObject();
+
```

Comment:
For clarity, might be good to put a comment here on what the message looks like fully formatted.